File: rxlifecycle/src/main/java/com/uniquext/android/rxlifecycle/RxLifecycle.java
package com.uniquext.android.rxlifecycle;

import com.uniquext.android.rxlifecycle.event.ActivityEvent;
import com.uniquext.android.rxlifecycle.event.FragmentEvent;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;

/**
 * 　 　　   へ　　　 　／|
 * 　　    /＼7　　　 ∠＿/
 * 　     /　│　　 ／　／
 * 　    │　Z ＿,＜　／　　   /`ヽ
 * 　    │　　　 　　ヽ　    /　　〉
 * 　     Y　　　　　   `　  /　　/
 * 　    ｲ●　､　●　　⊂⊃〈　　/
 * 　    ()　 へ　　　　|　＼〈
 * 　　    >ｰ ､_　 ィ　 │ ／／      去吧！
 * 　     / へ　　 /　ﾉ＜| ＼＼        比卡丘~
 * 　     ヽ_ﾉ　　(_／　 │／／           消灭代码BUG
 * 　　    7　　　　　　　|／
 * 　　    ＞―r￣￣`ｰ―＿
 * ━━━━━━感觉萌萌哒━━━━━━
 *
 * @author UniqueXT
 * @version 1.0
 * @date 2019/4/25  17:05
 */
public class RxLifecycle {

    public static <Upstream> ObservableTransformer<Upstream, Upstream> bindUntilEvent(Observable<ActivityEvent> eventSubject, ActivityEvent event) {
        return bind(targetEventSubject(eventSubject, event));
    }

    public static <Upstream> ObservableTransformer<Upstream, Upstream> bindUntilEvent(Observable<FragmentEvent> eventSubject, FragmentEvent event) {
        return bind(targetEventSubject(eventSubject, event));
    }

    private static <Upstream, Event> LifecycleTransformer<Upstream> bind(ObservableSource<Event> eventSubject) {
        return new LifecycleTransformer<>(eventSubject);
    }

    private static ObservableSource<ActivityEvent> targetEventSubject(Observable<ActivityEvent> eventSubject, final ActivityEvent event) {
        return eventSubject.filter(e -> e.ordinal() - event.ordinal() >= 0);
    }

    private static ObservableSource<FragmentEvent> targetEventSubject(Observable<FragmentEvent> eventSubject, final FragmentEvent event) {
        return eventSubject.filter(e -> e.ordinal() - event.ordinal() >= 0);
    }
}

File: rxlifecycle/src/main/java/com/uniquext/android/rxlifecycle/RxLifecycle.kt
package com.uniquext.android.rxlifecycle

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import com.uniquext.android.rxlifecycle.transformer.NullTransformer
import io.reactivex.ObservableTransformer
import java.util.concurrent.ConcurrentHashMap

object RxLifecycle {

    private val lifecycleMap: ConcurrentHashMap<LifecycleOwner, RxLifecycleObserver> = ConcurrentHashMap()

    @JvmStatic
    fun <Upstream> untilCreate(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_CREATE)
        }
    }

    @JvmStatic
    fun <Upstream> untilStart(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_START)
        }
    }

    @JvmStatic
    fun <Upstream> untilResume(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_RESUME)
        }
    }

    @JvmStatic
    fun <Upstream> untilPause(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_PAUSE)
        }
    }

    @JvmStatic
    fun <Upstream> untilStop(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_STOP)
        }
    }

    @JvmStatic
    fun <Upstream> untilDestroy(lifecycleOwner: LifecycleOwner): ObservableTransformer<Upstream, Upstream> {
        return if (lifecycleMap[lifecycleOwner] == null) {
            NullTransformer()
        } else {
            lifecycleMap[lifecycleOwner]!!.bindUntilEvent(Lifecycle.Event.ON_DESTROY)
        }
    }

    @JvmStatic
    fun bind(lifecycleOwner: LifecycleOwner) {
        if(lifecycleMap[lifecycleOwner] == null) {
            val lifecycleObserver = RxLifecycleObserver()
            lifecycleMap[lifecycleOwner] = lifecycleObserver
            lifecycleOwner.lifecycle.addObserver(lifecycleObserver)
        }
    }

    @JvmStatic
    fun unbind(lifecycleOwner: LifecycleOwner) {
        val lifecycleObserver = lifecycleMap.remove(lifecycleOwner)
        if (lifecycleObserver != null) {
            lifecycleOwner.lifecycle.removeObserver(lifecycleObserver)
        }
    }
}